Version 4.1.0

skipperbent opened the pull request:
- Removed `usmanhalalit/viocon` composer dependency (Container support) to increase library performance.
- Moved event-types constants to `EventHandler` from `QueryBuilderHandler`.
- Added return-types to more methods.
- Added `IQueryBuilderHandler` interface for more implementations of return-types.
- Added more return-types to methods.
- Fixed return-type for some phpDoc blocks and methods.
- Fixed insert so it uses Transactions when inserting multiple/batch.
- Fixed issue where bindings with null-value was unavailable/missing from query  (issue: #35).
- Added unit tests for bindings with null-value (issue: #35)
- Changed PDO-parameter behavior so it automatically parses correct parameter-type.
- Simplified PDO-parameter bindings on queries.
- Transactions are now working correctly for batch-insert (issue: #31).
- Fixed nested transactions not working and added unit-tests (issue: #31).
- Added very basic unit-tests for Transactions.
- Minor optimisation, cleanup and bug-fixes.
- Updated existing unit-tests.
- Updated documentation.

## Release note
This version contains breaking changes. Before upgrading, please read the read the release notes carefully.

**Removed `usmanhalalit/viocon` composer dependency (Container support) to increase library performance.**
If you are currently using containers in your projects, please note that this feature is no longer supported in the `pecee-pixie` library. To create custom connection-adapter, you need to inherit the `IConnectionAdapter` interface and use it as a driver on your `Connection` object instead.

**Moved event-type constant from `QueryBuilderHandler` to `EventHandler` class.**
Event-type constant has been moved to the `EventHandler` class. Please change all references to `QueryBuilderHandler::EVENT_BEFORE_UPDATE` etc. to `EventHandler::EVENT_BEFORE_UPDATE`.

skipperbent committed:
Development
- Added IQueryBuilderHandler interface for more implementations of return-types.
- Added more return-types to methods.
- Fixed return-type for some phpDoc blocks.
- Changed batch-insert so it uses PDO transactions.

skipperbent commented:
@mrcnpdlk Do you have any comments on the batch-insert transaction functionality? I was unable to find a way to test it using php-unit.

mrcnpdlk commented:
Unfortunately it doesn't work.
Transaction class object (`$qb`) inside callback function doesn't able to fire `getQuery($type, $subData)` because Transaction object has empty required params.

In my opinion the best solution is remove Transaction class and move method from it to QueryBuilderHandler. After that, those methods will be available without create Transaction class object.

skipperbent commented:
@mrcnpdlk great idea, I'll implement something like that instead. Thanks for your feedback 👍👌

skipperbent committed:
[WIP!!!] Development
- Fixed issue where bindings with null-value was removed (issue: #35).
- Added unit tests for bindings with null-value (issue: #35)
- Transactions are now fully working.
- Added unit-tests for Transactions.
- Changed PDO-parameter behavior so it automaticially parses correct parameter-type.
- Began implementing Transactions for batch-insert, update etc (issue: #31).

skipperbent committed:
Transactions are now working correctly for batch-insert.

skipperbent committed:
Fixed nested transactions not working and added unit-tests.

skipperbent committed:
Updated documentation

skipperbent committed:
Development

- Removed `usmanhalalit/viocon` composer dependency (Container support) to increase library performance.
- Moved event-types constants to `EventHandler` from `QueryBuilderHandler`.
- Added return-types to more methods.
- Updated documentation.
- Updated unit-tests to reflect new changes.
